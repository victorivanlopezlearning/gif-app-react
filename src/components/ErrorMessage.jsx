
const ErrorMessage = ({ errorMesagge }) => {
  return (
    <div className="error">
      <p>{errorMesagge}</p>
    </div>
  )
}

export default ErrorMessage;