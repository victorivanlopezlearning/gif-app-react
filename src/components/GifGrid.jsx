import { useEffect, useState } from "react";
import { getGifs } from "../helpers";

const GifGrid = ({ category }) => {

  const [imagesGifs, setImagesGifs] = useState([]);

  const getImagesFromGifs = async () => {
    const images = await getGifs(category);
    setImagesGifs(images);
  };

  useEffect(() => {
    getImagesFromGifs();
  }, [])

  return (
    <>
      <h2>{category}</h2>

      <ol>
        {imagesGifs.map(({ id, title }) => (
          <li key={id}>{title}</li>
        ))}
      </ol>
    </>
  )
}

export default GifGrid;