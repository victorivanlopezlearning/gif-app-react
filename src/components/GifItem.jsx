
const GifItem = ({ title, url }) => {

  return (
    <div className="card">
      <img src={url} alt={`Imagen ${title}`} />
      <div className="card-body">
        <p>{title}</p>
      </div>
    </div>
  )
}

export default GifItem;