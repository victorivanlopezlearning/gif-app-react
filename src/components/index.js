
export * from './ErrorMessage';
export * from './Form';
export * from './GifGrid';
export * from './GifItem';