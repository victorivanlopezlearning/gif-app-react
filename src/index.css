html {
  box-sizing: border-box;
  font-size: 62.5%;
}

*,
*:before,
*:after {
  box-sizing: inherit;
}

body {
  font-family: Helvetica, Arial, sans-serif;
  max-width: 1200px;
  margin: 0 auto;
  width: 90%;
  font-size: 2rem;
  background-color: #f3eae8;
}

input {
  background-color: white;
  border-radius: 5px;
  border: 1px solid #f6aca1;
  color: #000000;
  font-size: 1.5rem;
  outline: none;
  padding: 1.5rem;
  width: 100%;
}

h1,
h2,
h3 {
  margin: 2rem 0;
}

h1 {
  font-size: 4rem;
  text-align: center;
}

h2 {
  font-size: 3rem;
}

h3 {
  font-size: 2rem;
  margin-bottom: 5px;
}

.error {
  padding: 1rem;
}
.error p {
  margin: 0;
  font-size: 1.5rem;
  font-weight: bold;
  color: #921212;
}

.card-grid {
  display: grid;
  grid-template-columns: 1fr;
  gap: 2rem;
  margin-bottom: 2rem;
}
@media (min-width: 768px) {
  .card-grid {
    display: grid;
    grid-template-columns: repeat(3, 1fr);
  }
}

.card {
  align-items: center;
  background-color: white;
  border-radius: 10px;
  box-shadow: 0px 3px 5px rgba(0, 0, 0, 0.03);
  display: flex;
  flex-direction: column;
  justify-content: center;
  overflow: hidden;
}

.card p {
  background-color: white;
  font-size: 1.5rem;
  margin-top: 5px;
  padding: 5px 20px 0px 20px;
  text-align: center;
}

.card img {
  width: 40rem;
  height: 30rem;
  object-fit: cover;
}