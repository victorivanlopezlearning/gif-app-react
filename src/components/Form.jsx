import PropTypes from 'prop-types';
import { useState } from 'react';
import { ErrorMessage } from './ErrorMessage';

export const Form = ({ onAddCategory }) => {

  const [category, setCategory] = useState('');
  const [errorMessage, setErrorMessage] = useState('');

  const onSubmit = (e) => {
    e.preventDefault();

    const newCategory = category.trim();
    if ([newCategory].includes('')) {
      setErrorMessage('No debe ir vac√≠o');
      return;
    };
    setErrorMessage('');
    onAddCategory(newCategory);
    setCategory('');
  };

  return (
    <>
      <form
        aria-label='form'
        onSubmit={onSubmit}
      >
        <input
          type="text"
          placeholder="Buscar GIFs"
          onChange={({ target }) => setCategory(target.value)}
          value={category}
        />
      </form>
      {errorMessage && <ErrorMessage errorMessage={errorMessage} />}
    </>
  )
}

Form.propTypes = {
  onAddCategory: PropTypes.func.isRequired,
};