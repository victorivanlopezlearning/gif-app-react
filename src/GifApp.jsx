import { useState } from 'react';
import Form from './components/Form';
import ErrorMessage from './components/ErrorMessage';

const GifApp = () => {

  const [categories, setCategories] = useState([]);
  const [errorMesagge, setErrorMesagge] = useState('');

  const onAddCategory = (newCategory) => {

    const categoryExist = categories.find((category) => category.toUpperCase() === newCategory.toUpperCase());
    if (categoryExist) {
      setErrorMesagge(`${newCategory} ya est√° listado. Por favor, indica uno diferente.`);
      return;
    }
    setErrorMesagge('');
    setCategories([newCategory, ...categories]);
  };

  return (
    <>
      <h1>Buscador de GIFs</h1>

      <Form
        onAddCategory={onAddCategory}
      />
      {errorMesagge && <ErrorMessage errorMesagge={errorMesagge} />}

      <ol>
        {categories?.map(category => (
          <li key={category}>{category}</li>
        ))}
      </ol>
    </>
  )
}

export default GifApp;