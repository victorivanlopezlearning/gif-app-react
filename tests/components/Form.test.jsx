import { fireEvent, render, screen } from '@testing-library/react';
import { Form } from '../../src/components';

describe('Tests in <Form />', () => {
  test('Should match with the snapshot', () => {
    const { container } = render(<Form onAddCategory={() => { }} />);
    expect(container).toMatchSnapshot();
  });

  test('Should change the input value', () => {
    render(<Form onAddCategory={() => { }} />);

    const input = screen.getByRole('textbox');

    fireEvent.input(input, { target: { value: 'Metal Gear Solid' } });
    expect(input.value).toBe('Metal Gear Solid');
  });

  test('should call onAddCategory if the input has a value', () => {

    const inputValue = 'Metal Gear Solid';
    const onAddCategory = jest.fn(); // jest.fn() | es un Mock de simulación de la función

    render(<Form onAddCategory={onAddCategory} />);

    const input = screen.getByRole('textbox');
    const form = screen.getByRole('form');

    fireEvent.input(input, { target: { value: inputValue } });
    fireEvent.submit(form);

    expect(input.value).toBe('');
    expect(onAddCategory).toHaveBeenCalled(); // Valida que la función indicada fue llamada al menos 1 vez
    expect(onAddCategory).toHaveBeenCalledTimes(1); // Valida que la función indicada fue llamada las veces que indiquemos
    expect(onAddCategory).toHaveBeenCalledWith(inputValue); // Valida que la función indicada fue mandada a llamar con el valor del inputValue 
  });

  test('Should show the error if the input is empty', () => {

    const onAddCategory = jest.fn();

    render(<Form onAddCategory={onAddCategory} />);

    const form = screen.getByRole('form');

    fireEvent.submit(form);

    expect(screen.getByText('No debe ir vacío')).toBeTruthy(); // Confirmar que el texto aparece en el HTML
    expect(onAddCategory).toHaveBeenCalledTimes(0); // La función no fue llamada
    expect(onAddCategory).not.toHaveBeenCalled(); // La función no fue llamada
  });
});