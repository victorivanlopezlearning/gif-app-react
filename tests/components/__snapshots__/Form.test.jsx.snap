// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <Form /> Should match with the snapshot 1`] = `
<div>
  <form
    aria-label="form"
  >
    <input
      placeholder="Buscar GIFs"
      type="text"
      value=""
    />
  </form>
</div>
`;
